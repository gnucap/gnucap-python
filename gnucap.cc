#include "gnucap.h"
#include "gnucap/u_lang.h"
#include "gnucap/c_comand.h"
#include "gnucap/globals.h"
#include "gnucap/m_wave.h"
#include "gnucap/s__.h"
#include "gnucap/io_.h"

#include "numpy_interface.h"

#include <stdio.h>
#include <string>
#include <fstream>

std::string command(char *command) {
  
  char filename[L_tmpnam];
  
  tmpnam(filename);
  
  // supress output to stdout
  IO::mstdout.detach(stdout);

  // send output to file
  CMD::command(std::string("> ") + std::string(filename), &CARD_LIST::card_list);

  CMD::command(std::string(command), &CARD_LIST::card_list);

  CMD::command(">", &CARD_LIST::card_list);

  // Open file an read it
  std::ifstream ifs(filename);

  std::ostringstream oss;

  oss << ifs.rdbuf();

  std::string output(oss.str());

  unlink(filename);
  
  return output;
}

DISPATCHER<CMD>::INSTALL *attach_command(char *command_name, CMD *cmd) {
  return new DISPATCHER<CMD>::INSTALL(&command_dispatcher, command_name, cmd);
}

void detach_command(DISPATCHER<CMD>::INSTALL *installer) {
  delete installer;
}
