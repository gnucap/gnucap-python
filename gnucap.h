#include "gnucap/c_comand.h"
#include "gnucap/l_dispatcher.h"
#include "gnucap/s__.h"

class SIMWrapper : public SIM {
public:
  explicit SIMWrapper():SIM()  {}
  virtual void  setup(CS&)=0;
  virtual void  sweep()=0;
};

std::string command(char *command);
DISPATCHER<CMD>::INSTALL *attach_command(char *command_name, CMD *cmd);
void detach_command(DISPATCHER<CMD>::INSTALL *installer);

