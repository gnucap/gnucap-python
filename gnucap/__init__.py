# python3 only: use importlib

import sys
import ctypes
import os

# python tries to outsmart us, work around that
if(os.name=="posix"):
	flags = sys.getdlopenflags()
	sys.setdlopenflags(flags | ctypes.RTLD_GLOBAL)
	from gnucap_swig import *
	from component import component
	sys.setdlopenflags(flags)
else:
	untested()
	from gnucap_swig import *
	from component import component

# TODO: ask gnucap-conf (at configure time)
# BUG: do not override, if set.
os.environ["GNUCAP_PLUGPATH"] = "/usr/local/lib/gnucap"

if hasattr(sys, 'ps1'):
	print("welcome to gnucap-python")
	run_mode = SET_RUN_MODE(rINTERACTIVE)
else:
	run_mode = SET_RUN_MODE(rBATCH)
	if sys.flags.interactive:
		# what is this?
		print("... in interactive postmortem mode, incomplete")

try:
	lang = os.environ["GNUCAP_DEFAULT_PLUGINS"]
except:
	default_plugins = "gnucap-default-plugins.so"

try:
	lang = os.environ["GNUCAP_LANG"]
except:
	lang = "acs";

command("load " + default_plugins)
command("set lang=" + lang)

# this is the plan
# for s in simulations:
#	attach_output(s, our_own_sink)
