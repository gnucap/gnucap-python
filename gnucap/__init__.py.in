# python3 only: use importlib

import sys
import ctypes
import os

# python tries to outsmart us, work around that
if(os.name=="posix"):
	flags = sys.getdlopenflags()
	sys.setdlopenflags(flags | ctypes.RTLD_GLOBAL)
	from gnucap_swig import *
	sys.setdlopenflags(flags)
else:
	untested()
	from gnucap_swig import *

# trying to manifest an interface.
# ideally the type names should be the usual ones, wrapper layers must be invisible.
from component import COMPONENT_ as COMPONENT
from e_elemnt import ELEMENT
from gnucap_swig import CARD_LIST
from gnucap_swig import SIM_ as SIM
# from gnucap_swig import DPAIR
# from gnucap_swig import COMPLEX
from m_cpoly import FPOLY1, CPOLY1
from m_wave import WAVE
from e_node import node_t
from c_exp import eval

# this will change
from globals import install_device
from globals import install_command
# from globals import shared_command_installer as command_installer

# TODO: ask gnucap-conf (at configure time)
# BUG: do not override, if set.
os.environ["GNUCAP_PLUGPATH"] = "/usr/local/lib/gnucap"

if hasattr(sys, 'ps1'):
	print("welcome to gnucap-python")
	run_mode = SET_RUN_MODE(rINTERACTIVE)
else:
	run_mode = SET_RUN_MODE(rBATCH)
	if sys.flags.interactive:
		# what is this?
		print("... in interactive postmortem mode, incomplete")

try:
	lang = os.environ["GNUCAP_DEFAULT_PLUGINS"]
except:
	default_plugins = "gnucap-default-plugins.so"

try:
	lang = os.environ["GNUCAP_LANG"]
except:
	lang = "acs";

command("load " + default_plugins)
command("set lang=" + lang)

# this is the plan
# for s in simulations:
#	attach_output(s, our_own_sink)
