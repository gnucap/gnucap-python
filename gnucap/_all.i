/* Copyright (C) 2020 Felix Salfelder
 * Author: Felix Salfelder <felix@salfelder.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 *------------------------------------------------------------------
 */

%module all

%pythoncode %{
stub=None
%}

%include _ap.i
%include _gnucap_swig.i
%include _s__.i
%include _e_base.i
%include _e_card.i
%include _e_paramlist.i
%include _e_elemnt.i
%include _e_compon.i
%include _io_.i
%include _u_sim_data.i
%include _c_comand.i
%include _c_exp.i
%include director_except.i
%include _e_cardlist.i
%include _e_node.i
%include _e_subckt.i
%include _globals.i
%include _l_compar.i
%include _l_denoise.i
%include _m_cpoly.i
%include _md.i
%include _m_matrix.i
%include _mode.i
%include _m_wave.i
%include numpy.i
%include _u_function.i
%include _u_nodemap.i
%include _u_opt.i
%include _u_parameter.i
%include _u_status.i
%include _u_time_pair.i
%include _u_xprobe.i
