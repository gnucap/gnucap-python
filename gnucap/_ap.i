/* Copyright (C) 2018 Felix Salfelder
 * Author: Felix Salfelder <felix@salfelder.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 *------------------------------------------------------------------
 */
%module(directors="0", allprotected="1") ap

%{
#include <ap.h>
%}

%include std_string.i

enum AP_MOD{
  mNONE,       /* nothing special */
  mSCALE,      /* scale it after reading */
  mOFFSET,     /* add an offset */
  mINVERT,     /* save 1 / the number */
  mPOSITIVE,   /* store absolute value */
  mOCTAL,      /* read the number in octal */
  mHEX         /* read the number in hex */
};

%include "typemaps.i"

class CS {
public:
  enum STRING {_STRING};
  CS(CS::STRING, const std::string& s);

  // status - non-consuming
  unsigned cursor()const	{return _cnt;}
  bool	stuck(unsigned* INOUT)	{bool ok=*last<_cnt; *last=_cnt; return !ok;}
  bool	gotit(unsigned last)	{return last<_cnt;}
	operator bool()const	{return _ok;}

  // get -- non-consuming
  const std::string fullstring()const		{return _cmd;}
  const std::string substr(unsigned i)const {return ((_cmd.length()>=i) ? _cmd.substr(i) : "");}
  const std::string substr(unsigned i, unsigned n)const	{return _cmd.substr(i,n);}
  const std::string tail()const			{return substr(_cnt);}
  char		    peek()const			{return _cmd[_cnt];}

  // status - may consume whitespace only
  bool	      ns_more()const	{return peek()!='\0';}
  bool	      more()		{skipbl(); return ns_more();}
  bool	      is_end()		{return !more();}
  bool	      is_file()		{return (_file && !isatty(fileno(_file)));}
  bool	      is_first_read()const {untested(); return (_line_number == 0);}

  // control
  CS&	      reset(unsigned c=0) {_cnt=c; _ok=true; return *this;}

  // exception handling (ap_error.cc) non-consuming
  CS&	      check(int, const std::string&);
  CS&	      warn(int, unsigned, const std::string&);
  CS&         warn(int i, const std::string& s)	{return warn(i,cursor(), s);}

  // string matching (ap_match.cc) possibly consuming, sets _ok
  CS&	      umatch(const std::string&);
  CS&	      scan(const std::string&);
  std::string last_match()const;
  std::string trimmed_last_match(const std::string& = " ,=;")const;

  // character tests - non-consuming, no _ok
  bool	      match1(char c)const{return (peek()==c);}
  bool	      match1(const std::string& c)const
		{return ns_more() && strchr(c.c_str(),peek());}
  size_t      find1(const std::string& c)const
	{return ((ns_more()) ? c.find_first_of(peek()) : std::string::npos);}
  bool	      is_xdigit()const
		{untested(); return (match1("0123456789abcdefABCDEF"));}
  bool	      is_digit()const	{return (match1("0123456789"));}
  bool	      is_pfloat()const	{return (match1(".0123456789"));}
  bool	      is_float()const	{return (match1("+-.0123456789"));}
  bool	      is_argsym()const	{return (match1("*?$%_&@"));}
  bool	      is_alpha()const	{return !!isalpha(toascii(peek()));}
  bool	      is_alnum()const   {return !!isalnum(toascii(peek()));}
  bool	      is_term(const std::string& t = ",=(){};")
	{char c=peek(); return (c=='\0' || isspace(c) || match1(t));}


public:
  std::string ctos(const std::string& term=",=(){};",
		   const std::string& b="\"'{",
		   const std::string& e="\"'}",
		   const std::string& trap="");
//...
  double      ctof();
  bool	      ctob();
  int	      ctoi();
  unsigned    ctou();
  int	      ctoo();
  int	      ctox();
  double      ctopf()			 {return std::abs(ctof());}

  %rename(myshift) operator>>;
  CS&	      operator>>(bool& INOUT)	 {x=ctob();return *this;}
  CS&	      operator>>(char& INOUT)	 {untested(); x=ctoc();return *this;}
  CS&         operator>>(int& INOUT)	 {x=ctoi();return *this;}
  CS&         operator>>(unsigned& INOUT)	 {x=ctou();return *this;}
  CS&         operator>>(double& INOUT)	 {x=ctof();return *this;}
  CS&	      operator>>(std::string& INOUT) {x=ctos();return *this;}
}; // CS

%pythoncode %{
from .ap import CS
_CS_rshift = dict()
from .ap import _CS_rshift
_CS_rshift[int] = CS.myshift
%}

%extend CS{
%pythoncode %{
	def __repr__(self):
		return "CS("+self.fullstring()+")"
	def stuck_(self, L):
		a, b=self.stuck(L[0])
		L[0]=b
		return a
	def __rshift__(self, b):
		if isinstance(b, list):
			r, x = _CS_rshift[type(b[0])](self, b[0])
			b[0]=x
		else:
			r = _CS_rshift[type(b)](self, b)
		return r;

%}
}



// template <class T>
bool Set(CS& cmd, const std::string& key, bool* INOUT, bool newval);
bool Set(CS& cmd, const std::string& key, double* INOUT, double newval);
bool Set(CS& cmd, const std::string& key, int* INOUT, int newval);

%pythoncode %{
def mySet(cmd, s, L, *args):
  if(isinstance(L, list)):
    a, b = Set(cmd, s, L[0], *args)
    L[0] = b
    return a
  else:
    return Set(cmd, s, L, *args)

%}

/*--------------------------------------------------------------------------*/

//%include "typemaps.i"
bool Get(CS& cmd, const std::string&, bool* INOUT);
bool Get(CS& cmd, const std::string&, int* INOUT,    AP_MOD=mNONE, int=0);
bool Get(CS& cmd, const std::string&, double* INOUT, AP_MOD, double=0.);


%pythoncode %{

_getD = dict()
_getD[int] = Get
_getD[float] = Get
_getD[bool] = Get

def _chooseGet(*args):
   return _getD[type(args[2])](*args)

def myGet(cmd, s, L, *args):
  if(isinstance(L, list)):
    a, b = _chooseGet(cmd, s, L[0], *args)
    L[0] = b
    return a
  else:
    return _chooseGet(cmd, s, L, *args)
%}



%pythoncode %{
from .ap import CS
%}

// vim:ts=8:sw=2
