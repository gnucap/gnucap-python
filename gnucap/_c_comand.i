/* Copyright (C) 2018 Felix Salfelder
 * Author: Felix Salfelder <felix@salfelder.org>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3, or (at your option)
 * any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
 * 02110-1301, USA.
 *------------------------------------------------------------------
 */
%module(directors="0", allprotected="1") c_comand
%feature("director") CMD;

%pythoncode %{
from .io_trace import untested
%}

%include _e_base.i

%exception {
    try {
        $action
    } catch (Exception& e) {
      PyErr_SetString(PyExc_Exception, e.message().c_str());
    }
}
%allowexception;

%{
#include <c_comand.h>
%}

class CMD : public CKT_BASE {
protected:
  explicit CMD();
public:
  std::string value_name()const {return "";}
  virtual void do_it(CS&, CARD_LIST*) = 0;
  virtual ~CMD() {}
  static  void  cmdproc(CS&, CARD_LIST*);
  static  void	command(const std::string&, CARD_LIST*);
};

%pythoncode %{
try:
  from .c_comand import CMD
except ImportError:
  from .all import stub
%}

