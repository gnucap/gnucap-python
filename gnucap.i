%module(directors="0", allprotected="1") gnucap

// generate directors for all classes that have virtual methods
%feature("director");
//%feature("nodirector") TRANSIENT; 

%include stl.i
%include std_string.i
%include std_complex.i

%{
#include "gnucap.h"
#include "gnucap/ap.h"
#include "gnucap/c_comand.h"
#include "gnucap/l_dispatcher.h"
#include "gnucap/s__.h"
#include "gnucap/m_wave.h"
#include "gnucap/u_opt.h"
#include "gnucap/e_cardlist.h"
#include "gnucap/globals.h"
#include "gnucap/md.h"
#include "gnucap/m_matrix.h"
#include "gnucap/u_status.h"
#include "gnucap/s_tr.h"
%}

#ifdef HAS_NUMPY
%{
#include "numpy_interface.h"
%}
#endif

%exception {
    try {
        $action
    } catch (Exception& e) {
      PyErr_SetString(PyExc_Exception, e.message().c_str());
      return NULL;
    }
}
%allowexception;

///////////////////////////////////////////////////////////////////////////////
// Basic types
///////////////////////////////////////////////////////////////////////////////
//%template(COMPLEX) std::complex<double>;

///////////////////////////////////////////////////////////////////////////////
// BSMATRIX
///////////////////////////////////////////////////////////////////////////////
template<class T> class BSMATRIX {
public:
  BSMATRIX(int ss=0);

  void          iwant(int, int);
  void          unallocate();
  void          allocate();
  void          reallocate();
  int           size()const;
  double        density();
  void          zero();
  void          dezero(T& o);
  void          load_diagonal_point(int i, T value);
  void          load_point(int i, int j, T value);
  void          load_couple(int i, int j, T value);
  void          load_symmetric(int i, int j, T value);
  void          load_asymmetric(int r1, int r2, int c1, int c2, T value);

  void          lu_decomp(const BSMATRIX<T>&, bool do_partial);
  void          lu_decomp();
  void          fbsub(T* v) const;
  void          fbsub(T* x, const T* b, T* c = NULL) const;

  T     d(int r, int  )const    {return *(_diaptr[r]);}
  T&    m(int r, int c);

private:
  T& m(int r, int c);
};

%template(BSMATRIXd) BSMATRIX<double>;
%template(BSMATRIXc) BSMATRIX<COMPLEX>;


///////////////////////////////////////////////////////////////////////////////
// Major gnucap classes
///////////////////////////////////////////////////////////////////////////////

class CS {
public:
      enum STRING {_STRING};
      CS(CS::STRING, const std::string& s);
      const std::string fullstring()const;
};

class CARD_LIST {
public:
   CARD_LIST& expand();
   CARD_LIST& precalc();
   CARD_LIST& map_nodes();
   CARD_LIST& tr_iwant_matrix();
   CARD_LIST& tr_begin();
   CARD_LIST& tr_restore();
   CARD_LIST& dc_advance();
   CARD_LIST& tr_advance();
   CARD_LIST& tr_regress();
   bool       tr_needs_eval()const;
   CARD_LIST& tr_queue_eval();
   bool       do_tr();
   CARD_LIST& tr_load();
   TIME_PAIR  tr_review();
   CARD_LIST& tr_accept();
   CARD_LIST& tr_unload();
   CARD_LIST& ac_iwant_matrix();
   CARD_LIST& ac_begin();
   CARD_LIST& do_ac();
   CARD_LIST& ac_load();

   static CARD_LIST card_list; // in globals.cc
};

class CKT_BASE {
protected:
  explicit CKT_BASE()                     :_probes(0), _label() {}
  explicit CKT_BASE(const std::string& s) :_probes(0), _label(s) {}
  explicit CKT_BASE(const CKT_BASE& p)    :_probes(0), _label(p._label) {}
  virtual  ~CKT_BASE();
public:

  static BSMATRIX<double>  aa;  /* raw matrix for DC & tran */
  static BSMATRIX<double>  lu;  /* decomposed matrix for DC & tran */
  static BSMATRIX<COMPLEX> acx; /* raw & decomposed matrix for AC */



protected:
  void set_command_ac()const      {_mode = s_AC;}
  void set_command_dc()const      {_mode = s_DC;}
  void set_command_op()const      {_mode = s_OP;}
  void set_command_tran()const    {_mode = s_TRAN;}
  void set_command_fourier()const {_mode = s_FOURIER;}

};

class CARD : public CKT_BASE {
protected:                              // create and destroy.
  CARD();
  CARD(const CARD&);
public:
  virtual  ~CARD()                      {delete _subckt;}

public: // parameters
  virtual std::string value_name()const = 0;

  virtual bool param_exists(int i)const {return param_name(i) != "";}
  virtual bool param_is_printable(int)const;
  virtual std::string param_name(int)const;
  virtual std::string param_name(int,int)const;
  virtual std::string param_value(int)const;
  virtual void set_param_by_name(std::string, std::string);
  virtual void set_param_by_index(int, std::string&, int);
  virtual int param_count()const {return 0;}
};

class CMD : public CARD { 
public:            
      CMD();
      virtual ~CMD();
      virtual void do_it(CS& cmd, CARD_LIST*)=0;
      std::string value_name()const;
      static void command(const std::string&, CARD_LIST*);
      static void cmdproc(CS&, CARD_LIST*);
};

class SIM : public CMD {
protected:
         SIM();
public:
                ~SIM();

        static void   init();
        static void   uninit();
 
        static double freq;           /* AC frequency to analyze at (Hertz) */
        static std::complex<double> jomega;        /* AC frequency to analyze at (radians) */
        static double time0;          /* time now */
        static double time1;          /* time at previous time step */
        static double _dtmin;         /* min internal step size */
        static double temp_c_in;      /* ambient temperature, input and sweep variable */
        static double temp_c;         /* ambient temperature, actual */
        static double damp;           /* Newton-Raphson damping coefficient actual */
        static bool uic;              /* flag: use initial conditions (spice-like) */
        static bool bypass_ok;        /* flag: ok to bypass model evaluation */
        static bool fulldamp;         /* flag: big iter. jump. use full (min) damp */
        static bool limiting;         /* flag: node limiting */
        static bool freezetime;       /* flag: don't advance stored time */
        static double genout;         /* tr dc input to circuit (generator) */

        static double last_time;      /* time at which "volts" is valid */
        static int    *nm;            /* node map (external to internal)      */
        static double *i;             /* dc-tran current (i) vector           */
        static double *v0;            /* dc-tran voltage, new                 */
        static double *vt1;           /* dc-tran voltage, 1 time ago          */
                                /*  used to restore after rejected step */
        static double *fw;            /* dc-tran fwd sub intermediate values  */
        static double *vdc;           /* saved dc voltages                    */
        static COMPLEX *ac;           /* ac right side                        */

        static WAVE* find_wave(const std::string&);
private:
        virtual void  setup(CS&)      = 0;
        virtual void  sweep()         = 0;
        virtual void  finish()        {}
        virtual bool  is_step_rejected()const {return false;}
};

// The SIMWrapper is needed since Swig doesn't handle private virtual methods
// All non-status methods that are inherited from SIM should also be copied 
// here or you will get segmentation faults
class SIMWrapper : public SIM {
public:
  explicit SIMWrapper():SIM()  {}
  virtual void  setup(CS&)      = 0;
  virtual void  sweep()         = 0;
protected:
         const PROBELIST& alarmlist()const;     /* s__out.cc */
         const PROBELIST& plotlist()const;
         const PROBELIST& printlist()const;
         const PROBELIST& storelist()const;
         void   outdata(double);
         void   head(double,double,const std::string&);
         void   print_results(double);
         void   alarm();
         virtual void  store_results(double);
private:
        const std::string long_label()const {unreachable(); return "";}
protected: 
         void   alloc_vectors();
  static void   unalloc_vectors(); 

};

class TRANSIENT : public SIM {
public:
        void do_it(CS&, CARD_LIST* scope);
        TRANSIENT();
        ~TRANSIENT();
        virtual void accept();
private:
        void  setup(CS&);
protected:
        bool _cont;
        void sweep();
        void outdata(double);
};

class STATUS {
public:
//  void command(CS& cmd);

  int user_nodes;
  int subckt_nodes;
  int model_nodes;
  int total_nodes;
  int control;
  int hidden_steps;
  int iter[iCOUNT];
};

enum RUN_MODE {
  rPRE_MAIN,    /* it hasn't got to main yet                    */
  rPRESET,      /* do set up commands now, but not simulation   */
                /* store parameters, so bare invocation of a    */
                /* simulation command will do it this way.      */
  rINTERACTIVE, /* run the commands, interactively              */
  rSCRIPT,      /* execute now, as a command, then restore mode */
  rBATCH        /* execute now, as a command, then exit         */
};

class SET_RUN_MODE {
public:
      SET_RUN_MODE(RUN_MODE rm);
};

///////////////////////////////////////////////////////////////////////////////
// Global variables
///////////////////////////////////////////////////////////////////////////////
//RUN_MODE ENV::run_mode = rPRE_MAIN;
DISPATCHER<CMD> command_dispatcher;
//DISPATCHER<COMMON_COMPONENT> bm_dispatcher;
//DISPATCHER<MODEL_CARD> model_dispatcher;
//DISPATCHER<CARD> device_dispatcher;
//DISPATCHER<LANGUAGE> language_dispatcher;
//DISPATCHER<FUNCTION> function_dispatcher;
STATUS status;

///////////////////////////////////////////////////////////////////////////////
// gnucap functions
///////////////////////////////////////////////////////////////////////////////
void command(char *command);
DISPATCHER<CMD>::INSTALL *attach_command(char *command_name, CMD *cmd);
void detach_command(DISPATCHER<CMD>::INSTALL *installer);

///////////////////////////////////////////////////////////////////////////////
// non-gnucap utility functions
///////////////////////////////////////////////////////////////////////////////
#ifdef HAS_NUMPY
void init_numpy();
PyObject *wave_to_arrays(WAVE *wave);
PyObject *bsmatrix_to_array_d(BSMATRIX<double> &A);
PyObject *bsmatrix_to_array_c(BSMATRIX<COMPLEX> &A);
PyObject *to_double_array(double *data, int len);
PyObject *get_complex_array(COMPLEX *data, int len);
void set_complex_array(COMPLEX *data, PyObject *srcarray);

template<class T> void bsmatrix_fbsub_array(BSMATRIX<T> *A, PyObject *rhs, PyObject *dest);

%template(bsmatrix_fbsub_array_double) bsmatrix_fbsub_array<double>;

#endif

///////////////////////////////////////////////////////////////////////////////
// init
///////////////////////////////////////////////////////////////////////////////
#ifdef HAS_NUMPY
%init %{
      init_numpy();
%}
#endif

///////////////////////////////////////////////////////////////////////////////
// python code
///////////////////////////////////////////////////////////////////////////////
//%pythoncode %{
//%}