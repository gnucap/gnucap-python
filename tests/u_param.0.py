
from gnucap import PARAMETERi, PARAMETERd, Get_, CS, Get

key="k"

ppp = PARAMETERd()
print("ppp", ppp)
print("gotit.", type(ppp))

print("====3b")
Cmd = CS(CS._STRING, "k=true")

a = False
L = [a]
r = Get_(Cmd, key, L)
assert(r)
assert(L[0]==True)


print("====bool")
Cmd = CS(CS._STRING, "k=0")

a = True
r = Get_(Cmd, key, a)
print(r)

print("====int")
Cmd = CS(CS._STRING, "k=1")

a = 2
r = Get_(Cmd, key, a)
print(r)

print("====intlist")
Cmd = CS(CS._STRING, "k=0")

a = 2
L = [a]
r = Get_(Cmd, key, L)
print(r, L)
assert(L[0]==0)



print("====0")

pd = PARAMETERd(0.)
assert(pd==0)
Cmd = CS(CS._STRING, "k=1.")
r, s=Get_(Cmd, key, pd)
assert(pd==1)

print(pd<=3.)

print("====")
pd = PARAMETERd()
L=[pd]
Cmd = CS(CS._STRING, "k=1.")
r=Get_(Cmd, key, L)
assert(r)
assert(L[0]==1)

print(pd<=3.)
print("====2")

a = PARAMETERi()
print(a>3)
Cmd = CS(CS._STRING, "k=1")

r = Get_(Cmd, "k", a)
print("?", r)

print("====4")

from gnucap import mNONE
L=[3.]
Cmd = CS(CS._STRING, "k=5.")
r = Get_(Cmd, key, L, mNONE)
assert(r)
assert(L[0]==5.)

print("====5")
Cmd = CS(CS._STRING, "k=-.1")

p = PARAMETERd()
sip=[p]

print(Cmd.tail(), "....")
r = Get_(Cmd, "k", sip)
print(r)
print(sip[0])
