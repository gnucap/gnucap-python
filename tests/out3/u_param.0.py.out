generator: already installed, replacing
stashing as generator:0
transient: already installed, replacing
stashing as transient:0
transient: already installed, replacing
stashing as transient:0
default plugins: master 2017.10.03
ppp NA( NA)
gotit. <class 'gnucap.u_parameter.PARAMETERd'>
====3b
====bool
[True, False]
====int
[True, 1]
====intlist
True [0]
====0
True
====
True
====2
False
? [True, 1]
====4
====5
k=-.1 ....
True
-0.1
